### PR TITLE
Simplify global search and library name

### PR DESCRIPTION
**Pull request by Siedlerchr**
Fixes parts of #8054 

Rely on Observable Magic and collect entries for each db context

Bildschirmfoto 2021-09-07 um 20 21 48

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java [hunk 1]
```diff
@@ -116,10 +119,13 @@ public BibEntry getEntry() {
...
-                observables.toArray(Observable[]::new));
+        value = Bindings.createStringBinding(() -> fieldValueFormatter.getValue().formatFieldsValues(fields, entry),
+                                             observables.toArray(Observable[]::new));
         fieldValues.put(fields, value);
         return value;
     }
+
+    public ObservableValue<String> getBibDatabaseContextPath() {
+        return new ReadOnlyStringWrapper(bibDatabaseContext.getDatabasePath().map(Path::toString).orElse(""));
+    }
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:130]: Not sure if this is a good idea, to make this an ObservableValue, which implies that it can be changed from the outside view on the signature of the method. Convention was in other cases to use Properties (in this case ReadOnlyStringProperty)

File: src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java [hunk 2]
```diff
@@ -71,13 +76,11 @@ public BibEntry getEntry() {
...
-                                             .map(GroupTreeNode::getGroup)
-                                             .filter(Predicate.not(Predicate.isEqual(groupTreeNode.getGroup())))
-                                             .collect(Collectors.toList()))
-                                .orElse(Collections.emptyList())));
+                                                      (a, b) -> database.getMetaData().getGroups().map(groupTreeNode -> groupTreeNode.getMatchingGroups(entry).stream()
+                                                                                                                                     .map(GroupTreeNode::getGroup)
+                                                                                                                                     .filter(Predicate.not(Predicate.isEqual(groupTreeNode.getGroup())))
+                                                                                                                                     .collect(Collectors.toList()))
+                                                                        .orElse(Collections.emptyList())));
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:83]: Whitespacechanges makes it way more harder to read

File: src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java [hunk 3]
```diff
@@ -47,22 +51,23 @@ public BibEntryTableViewModel(BibEntry entry, BibDatabaseContext bibDatabaseCont
...
+                                entry.getFieldBinding(StandardField.URI),
+                                entry.getFieldBinding(StandardField.EPRINT),
+                                (url, doi, uri, eprint) -> {
+                                    Map<Field, String> identifiers = new HashMap<>();
+                                    url.ifPresent(value -> identifiers.put(StandardField.URL, value));
+                                    doi.ifPresent(value -> identifiers.put(StandardField.DOI, value));
+                                    uri.ifPresent(value -> identifiers.put(StandardField.URI, value));
+                                    eprint.ifPresent(value -> identifiers.put(StandardField.EPRINT, value));
+                                    return identifiers;
+                                });
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:70]: unnecessary whitespace change

File: src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    private final SortedList<BibEntryTableViewModel> entriesSorted;
+    private final ObjectProperty<MainTableFieldValueFormatter> fieldValueFormatter;
+    private final PreferencesService preferencesService;
+    private final BibDatabaseContext bibDatabaseContext;
+
+    public SearchResultsTableDataModel(BibDatabaseContext bibDatabaseContext, PreferencesService preferencesService, StateManager stateManager) {
+        this.preferencesService = preferencesService;
+        this.bibDatabaseContext = bibDatabaseContext;
+        this.fieldValueFormatter = new SimpleObjectProperty<>(
+                                                              new MainTableFieldValueFormatter(preferencesService, bibDatabaseContext));
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java:40]: codestyle seems broken?

File: src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java [hunk 5]
```diff
@@ -82,13 +83,7 @@ private void initialize() {
         EasyBind.subscribe(viewModel.keepOnTop(), value -> {
             Stage stage = (Stage) getDialogPane().getScene().getWindow();
             stage.setAlwaysOnTop(value);
-            keepOnTop.setGraphic(value
-                    ? IconTheme.JabRefIcons.KEEP_ON_TOP.getGraphicNode()
-                    : IconTheme.JabRefIcons.KEEP_ON_TOP_OFF.getGraphicNode());
+            keepOnTop.setGraphic(value ? IconTheme.JabRefIcons.KEEP_ON_TOP.getGraphicNode() : IconTheme.JabRefIcons.KEEP_ON_TOP_OFF.getGraphicNode());
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java:86]: everything in one line makes it hard to read. Maybe a complete if clause?

File: src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java [hunk 6]
```diff
@@ -53,17 +52,19 @@ public GlobalSearchResultDialog(ExternalFileTypes externalFileTypes, UndoManager
...
     private void initialize() {
-        viewModel = new GlobalSearchResultDialogViewModel(stateManager, preferencesService);
+        viewModel = new GlobalSearchResultDialogViewModel(preferencesService);
 
         PreviewViewer previewViewer = new PreviewViewer(viewModel.getSearchDatabaseContext(), dialogService, stateManager);
         previewViewer.setTheme(preferencesService.getTheme());
         previewViewer.setLayout(preferencesService.getPreviewPreferences().getCurrentPreviewStyle());
 
-        FieldColumn fieldColumn = new FieldColumn(MainTableColumnModel.parse("field:library"));
+        //TODO: Create a special column
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java:61]: Why ToDo?

File: src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java [hunk 7]
```diff
@@ -53,17 +52,19 @@ public GlobalSearchResultDialog(ExternalFileTypes externalFileTypes, UndoManager
...
+        FieldColumn fieldColumn = new FieldColumn(MainTableColumnModel.parse("customLibrary"));
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java:62]: Instead of creating a special MainTableColumn, I would just use a plain old TableColumn if you just bind a static value later, should work here too.

File: src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        }
+        ObservableList<BibEntryTableViewModel> entriesViewModel = FXCollections.observableArrayList(models);
+
+        entriesFiltered = new FilteredList<>(entriesViewModel);
+        entriesFiltered.predicateProperty().bind(
+                                                 EasyBind.combine(stateManager.activeGroupProperty(), stateManager.activeSearchQueryProperty(), (groups, query) -> entry -> isMatched(groups, query, entry)));
+
+        IntegerProperty resultSize = new SimpleIntegerProperty();
+        resultSize.bind(Bindings.size(entriesFiltered));
+        stateManager.setActiveSearchResultSize(bibDatabaseContext, resultSize);
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java:57]: This seems to be the job the the GlobalSearchBar, not the SearchDialog!

File: src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                                                              new MainTableFieldValueFormatter(preferencesService, bibDatabaseContext));
+
+        List<BibEntryTableViewModel> models = new ArrayList<>();
+        for (BibDatabaseContext bibdatabasecontext : stateManager.getOpenDatabases()) {
+            ObservableList<BibEntry> entriesForDb = bibdatabasecontext.getDatabase().getEntries();
+            List<BibEntryTableViewModel> viewModelForDb = EasyBind.mapBacked(entriesForDb, entry -> new BibEntryTableViewModel(entry, bibdatabasecontext, fieldValueFormatter));
+
+            models.addAll(viewModelForDb);
+        }
+        ObservableList<BibEntryTableViewModel> entriesViewModel = FXCollections.observableArrayList(models);
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java:49]: Why not making the models list an ObservableList from the start?

File: src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+        entriesFiltered = new FilteredList<>(entriesViewModel);
+        entriesFiltered.predicateProperty().bind(
+                                                 EasyBind.combine(stateManager.activeGroupProperty(), stateManager.activeSearchQueryProperty(), (groups, query) -> entry -> isMatched(groups, query, entry)));
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java:53]: failed line break?

**Review by calixtus — CHANGES_REQUESTED** [review]
Some nitpicks, some questions.

File: src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java (hunk 1 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:130]: It's also only a static string, or I'm missing something?

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:130]: Yes, it's used in the Cell(Value)Factory for the column and therefore needs to be wrapped in a property 
I will fix the codestyle formatting later.

Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:130]: Then just use https://github.com/JabRef/jabref/blob/bbb39a8ffb3b866c2081abfbd43dfce1da73ac1f/src/main/java/org/jabref/gui/util/BindingsHelper.java#L135 in the cell factory.

Comment (calixtus) [reply src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:130]: I disagree.
It's not a static string if the name of the library can change (e.g. save as) the string changes. Convention was until now to provide those variables as properties, not as observables.

File: src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java (hunk 9 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/search/SearchResultsTableDataModel.java]: Suprisingly it works now. Previously I was getting an exception on calling addAll on the observableArraylist...

File: src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java (hunk 6 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/search/GlobalSearchResultDialog.java]: Obsolete now

File: src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java (hunk 1 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java]: I added this now as a property

**Review by calixtus — APPROVED** [review]
